src/backend/distributed/executor/repartition_logic.c

```diff
@@ -0,0 +1,252 @@
+#include "distributed/worker_manager.h"
+#include "distributed/multi_server_executor.h"
+#include "distributed/repartition_logic.h"
+#include "distributed/worker_transaction.h"
+#include "distributed/worker_manager.h"
+#include "distributed/transaction_management.h"
+#include "distributed/multi_task_tracker_executor.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/transmit.h"
+
+
+static void FillTaskGroups(List **allTasks, List **mergeTasks);
+static List * CreateTemporarySchemas(List *mergeTasks);
+static List * CreateJobIds(List *mergeTasks);
+static void CreateSchemasOnAllWorkers(char *createSchemasCommand);
+static char * GenerateCreateSchemasCommand(List *jobIds);
+static char * GenerateCommand(List *jobIds, char *templateCommand);
+static bool DoesJobIDExist(List *jobIds, uint64 jobId);
+static void SendCommandToAllWorkers(List *commandList);
+static char * GenerateDeleteJobsCommand(List *jobIds);
+static void RemoveTempJobDirs(List *jobIds);
+
+
+/*
+ * ExecuteDependedTasks executes all tasks except the top level tasks
+ * in order from the task tree. At a time, it can execute different tasks from
+ * different jobs.
+ */
+void
+ExecuteDependedTasks(List *topLevelTasks)
+{
+	List *mergeTasks = NIL;
+
+
+	EnsureNoModificationsHaveBeenDone();
+
+	List *allTasks = TaskAndExecutionList(topLevelTasks);
+
+	FillTaskGroups(&allTasks, &mergeTasks);
+
+	List *jobIds = CreateTemporarySchemas(mergeTasks);
+
+	ExecuteTasksInDependencyOrder(allTasks, topLevelTasks);
+
+	RemoveTempJobDirs(jobIds);
+}
+
+
+/*
+ * FillTaskGroups iterates all tasks and creates a group for mergeTasks.
+ */
+static void
+FillTaskGroups(List **allTasks, List **mergeTasks)
+{
+	ListCell *taskCell = NULL;
+
+	foreach(taskCell, *allTasks)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+
+		if (task->taskType == MERGE_TASK)
+		{
+			*mergeTasks = lappend(*mergeTasks, task);
+		}
+	}
+}
+
+
+/*
+ * CreateTemporarySchemas creates the necessary schemas that will be used
+ * later in each worker. Single transaction is used to create the schemas.
+ */
+static List *
+CreateTemporarySchemas(List *mergeTasks)
+{
+	List *jobIds = CreateJobIds(mergeTasks);
+	char *createSchemasCommand = GenerateCreateSchemasCommand(jobIds);
+	CreateSchemasOnAllWorkers(createSchemasCommand);
+	return jobIds;
+}
+
+
+/*
+ * CreateJobIds returns a list of unique job ids that will be used
+ * in mergeTasks.
+ */
+static List *
+CreateJobIds(List *mergeTasks)
+{
+	ListCell *taskCell = NULL;
+	List *jobIds = NIL;
+
+	foreach(taskCell, mergeTasks)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+		if (!DoesJobIDExist(jobIds, task->jobId))
+		{
+			jobIds = lappend(jobIds, (void *) task->jobId);
+		}
+	}
+	return jobIds;
+}
+
+
+/*
+ * CreateSchemasOnAllWorkers creates schemas in all workers.
+ */
+static void
+CreateSchemasOnAllWorkers(char *createSchemasCommand)
+{
+	List *commandList = list_make1(createSchemasCommand);
+
+	SendCommandToAllWorkers(commandList);
```

Comment:
The implementation on the function doesn't match its name. This seems to be an `SendCommandToAllWorkers` whereas `SendCommandToAllWorkers` should be `SendCommandListToAllWorkers`.